hpx/lcos/detail/sync_implementations.hpp

```diff
@@ -0,0 +1,141 @@
+        call(naming::id_type const& id, naming::address && addr, Ts &&... vs)
+        {
+            typedef typename Action::remote_result_type remote_result_type;
+
+            typedef traits::get_remote_result<Result, remote_result_type>
+                get_remote_result_type;
+
+            return get_remote_result_type::call(Action::execute_function(
+                addr.address_, addr.type_, std::forward<Ts>(vs)...));
+        }
+    };
+
+    template <typename Action>
+    struct sync_local_invoke_direct<Action, void>
+    {
+        template <typename ...Ts>
+        static void
+        call(naming::id_type const& id, naming::address && addr, Ts &&... vs)
+        {
+            Action::execute_function(
+                addr.address_, addr.type_, std::forward<Ts>(vs)...);
+        }
+    };
+
+    ///////////////////////////////////////////////////////////////////////////
+    template <typename Action, typename ...Ts>
+    typename hpx::traits::extract_action<Action>::type::local_result_type
+    sync_impl(launch policy, hpx::id_type const& id, Ts&&... vs)
+    {
+        typedef typename hpx::traits::extract_action<Action>::type action_type;
+        typedef typename action_type::local_result_type result_type;
+        typedef typename action_type::component_type component_type;
+
+        std::pair<bool, components::pinned_ptr> r;
+
+        naming::address addr;
+        if (agas::is_local_address_cached(id, addr) &&
+            can_invoke_locally<action_type>())
+        {
+            // route launch policy through component
+            policy = traits::action_select_direct_execution<Action>::call(
+                policy, addr.address_);
+
+            if (traits::component_supports_migration<component_type>::call())
+            {
+                r = traits::action_was_object_migrated<Action>::call(
+                        id, addr.address_);
+                if (!r.first)
+                {
+                    if (policy == launch::sync ||
+                        action_type::direct_execution::value)
+                    {
+                        return hpx::detail::sync_local_invoke_direct<
+                            action_type, result_type
+                        >::call(id, std::move(addr), std::forward<Ts>(vs)...);
+                    }
+                }
+            }
+            else if (policy == launch::sync ||
+                action_type::direct_execution::value)
+            {
+                return hpx::detail::sync_local_invoke_direct<
+                        action_type, result_type
+                    >::call(id, std::move(addr), std::forward<Ts>(vs)...);
+            }
+        }
+
+        return async_impl<Action>(policy, id, std::forward<Ts>(vs)...).get();
+    }
+
+    template <typename Action, typename ...Ts>
+    typename hpx::traits::extract_action<Action>::type::local_result_type
+    sync_impl(hpx::detail::sync_policy, hpx::id_type const& id, Ts&&... vs)
+    {
```

Comment:
Do we really need to duplicate the code here?

Reply:
Those two functions still look almost identical. Do we really need that here?

Reply:
Ok, ok - you got me there. I was hoping you wouldn't notice ;-) Should be fine now, though.